pkg/client/apiutil/restmapper.go

```diff
@@ -230,35 +260,43 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 	}
 
 	m.mapper = restmapper.NewDiscoveryRESTMapper(updatedGroupResources)
-	return nil
 }
 
-// findAPIGroupByNameLocked returns API group by its name.
-func (m *mapper) findAPIGroupByName(groupName string) (*metav1.APIGroup, error) {
-	// Looking in the cache first.
+// findAPIGroupByNameAndMaybeAggregatedDiscovery tries to finds the passed apiGroup.
+// If the server supports aggregated discovery, it will always perform that.
+func (m *mapper) findAPIGroupByNameAndMaybeAggregatedDiscovery(groupName string) (*metav1.APIGroup, bool, error) {
+	// Looking in the cache first
 	{
 		m.mu.RLock()
 		group, ok := m.apiGroups[groupName]
 		m.mu.RUnlock()
 		if ok {
-			return group, nil
+			return group, false, nil
 		}
 	}
 
 	// Update the cache if nothing was found.
-	apiGroups, err := m.client.ServerGroups()
+	apiGroups, maybeResources, failedGVs, err := m.client.GroupsAndMaybeResources()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get server groups: %w", err)
+		return nil, false, fmt.Errorf("failed to get server groups: %w", err)
 	}
 	if len(apiGroups.Groups) == 0 {
-		return nil, fmt.Errorf("received an empty API groups list")
+		return nil, false, fmt.Errorf("received an empty API groups list")
 	}
 
 	m.mu.Lock()
 	for i := range apiGroups.Groups {
 		group := &apiGroups.Groups[i]
 		m.apiGroups[group.Name] = group
 	}
+	for failedGV := range failedGVs {
+		delete(maybeResources, failedGV)
+	}
+	m.checkedIfSupportsAggregatedDiscovery.Store(true)
+	if len(maybeResources) > 0 {
+		m.supportsAggregatedDiscovery.Store(true)
```

Comment:
Why is it correct to check len(maybeResources) > 0 after failedGVs have been removed to infer that aggregated discovery is supported?

I would have assumed it is supported as soon as len(maybeResources) > 0 independent of the failedGVs

Reply:
Good point, updated